### PR TITLE
[RHCLOUD-47088][RHCLOUD-47089] Add MCP UPDATE and DELETE write tools

### PR DESCRIPTION
**Pull request by dsova06**
## Link(s) to Jira
- [RHCLOUD-47088](https://issues.redhat.com/browse/RHCLOUD-47088) -- MCP UPDATE tools
- [RHCLOUD-47089](https://issues.redhat.com/browse/RHCLOUD-47089) -- MCP DELETE tools

## Description of Intent of Change(s)

Extends the MCP tool layer with 15 write tools covering UPDATE and DELETE operations across all RBAC resource types.

**Infrastructure changes:**
- Extended `_clone_request` to support PUT, PATCH, and DELETE HTTP methods
- Extended `_call_view_write` to accept a `method` parameter (previously hardcoded to POST) and handle 204 No Content responses
- Added `_call_view_delete` helper that appends query parameters to the URL path via `urlencode()` -- Django's `RequestFactory.delete()` puts `data` in the body, not the query string, so explicit URL encoding is required for views that read `request.query_params`

**9 UPDATE tools (RHCLOUD-47088):**
- `update_group` (common) -- PUT full replacement by UUID or name
- `update_role_v1` (v1) -- PUT full replacement including permissions
- `patch_role_v1` (v1) -- PATCH partial update (name/display_name/description only)
- `update_role` (v2) -- PUT full replacement including permissions
- `update_role_binding` (v2) -- PUT set exact role list for a subject on a resource
- `update_workspace` (v2) -- PUT full replacement
- `move_workspace` (v2) -- POST move to new parent in hierarchy
- `update_cross_account_request` (common) -- PUT full replacement
- `patch_cross_account_request` (common) -- PATCH status change (approve/deny/cancel)

**6 DELETE tools (RHCLOUD-47089):**
- `delete_group` (common) -- delete a custom group
- `remove_principals_from_group` (common) -- remove users/service accounts
- `remove_roles_from_group` (v1) -- remove role associations
- `delete_role_v1` (v1) -- delete a custom role
- `bulk_delete_roles` (v2) -- atomically delete multiple roles
- `delete_workspace` (v2) -- delete a standard workspace

All DELETE tool descriptions start with "DESTRUCTIVE:" and state the operation is "IRREVERSIBLE".

`batch_delete_role_bindings` was intentionally skipped -- the ViewSet method is not implemented.

**Write gating:** All tools use `@register_tool(write=True)` and follow the existing three-level gating: `MCP_WRITE_ENABLED` flag, API version gating, Django view permission checks.

## Local Testing

```bash
# Run all MCP tests (255 tests)
pipenv run tox -e py312-fast -- tests.management.test_mcp_views

# Run only UPDATE tool tests
pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPUpdateGroupTests tests.management.test_mcp_views.MCPUpdateRoleV1Tests tests.management.test_mcp_views.MCPUpdateToolsV2Tests tests.management.test_mcp_views.MCPUpdateCrossAccountTests

# Run only DELETE tool tests
pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPDeleteGroupTests tests.management.test_mcp_views.MCPDeleteRoleV1Tests tests.management.test_mcp_views.MCPDeleteToolsV2Tests

# Full test suite (3335 tests)
pipenv run tox -e py312-fast
```

## Checklist
- [x] if API spec changes are required, is the spec updated? (no API spec changes -- MCP tools are internal)
- [x] are there any pre/post merge actions required? if so, document here. (none)
- [x] are theses changes covered by unit tests? (42 new tests across 7 test classes)
- [x] if warranted, are documentation changes accounted for? (ASCII reference tables updated in code)
- [x] does this require migration changes? (no)
- [ ] is there known, direct impact to dependent teams/components?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [x] Input Validation (all tools validate inputs before delegating to views; group resolution validates UUID/name)
- [x] Output Encoding (responses are JSON-encoded via `json.dumps`)
- [x] Authentication and Password Management (all tools use `requires_auth=True`; auth context forwarded via `_clone_request`)
- [x] Session Management (N/A -- stateless request delegation)
- [x] Access Control (three-level write gating; view permission classes enforce authorization)
- [x] Cryptographic Practices (N/A)
- [x] Error Handling and Logging (4xx/5xx wrapped in error structures; tool calls logged with org_id)
- [x] Data Protection (tenant isolation maintained through view delegation)
- [x] Communication Security (internal endpoint only -- `/_private/_a2s/mcp/`)
- [x] System Configuration (gated by `MCP_WRITE_ENABLED` flag)
- [x] Database Security (no direct queries -- all via Django ORM through ViewSet delegation)
- [x] File Management (N/A)
- [x] Memory Management (N/A)
- [x] General Coding Practices (follows existing MCP tool patterns; reviewed via /review-full pipeline)

[RHCLOUD-47088]: https://redhat.atlassian.net/browse/RHCLOUD-47088?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RHCLOUD-47089]: https://redhat.atlassian.net/browse/RHCLOUD-47089?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Extend MCP management views with update and delete write tools for RBAC resources and strengthen shared request/response handling for JSON-based write operations.

New Features:
- Add MCP write tools for updating groups, roles (v1 and v2), role bindings, workspaces, and cross-account requests.
- Add MCP write tools for deleting groups, group principals, group roles, v1 roles, v2 roles in bulk, and workspaces, all gated behind existing write controls.

Enhancements:
- Generalize internal view-calling helper to support GET/POST/PUT/PATCH/DELETE with JSON handling, query parameters, and 204 responses.
- Introduce shared group resolution helper for MCP tools to consistently resolve groups by UUID or name within tenant context.
- Update the internal MCP tool reference table to document new write tools and their corresponding API endpoints.

Tests:
- Add comprehensive test coverage for new UPDATE and DELETE MCP tools across v1, v2, and cross-account scenarios, including auth and error paths.
- Extend MCP write configuration tests to assert write gating and registration for all new tools.

### CODE REVIEW

**Comment by dsova06** [comment]
@sourcery-ai review

**Review by dsova06 — COMMENTED** [review]
The overall structure is solid. The refactored `_call_view_json` -> thin wrappers pattern is clean. `_resolve_group_for_tool` reduces boilerplate across 6 tools. Tool descriptions are consistent and the DESTRUCTIVE labels on DELETE tools are a nice touch. 42 tests covering auth, success paths, and edge cases.

### Would be good to clean up

1. **204 status label says "deleted" for all HTTP methods.** `_call_view_json` (mcp_views.py:332-333) converts all 204 responses to `{"status": "deleted"}`. Right now this only fires on DELETE operations since PUT/PATCH views return 200, so it's not a bug today. But using `{"status": "no_content"}` or gating on the method would make this more accurate if someone calls `_call_view_json` directly for a non-delete 204 in the future. Low effort change, and the existing test assertions for `output["status"] == "deleted"` would just need updating to match.

2. **`patch_cross_account_request` validation has no test.** The status validation added in `4835d6b3` (the `allowed_statuses` set at mcp_views.py:3736-3738) is good, but there's no test exercising the rejection path. A quick `test_patch_cross_account_request_invalid_status` calling with `status="foo"` and asserting the error mentions allowed statuses would lock this down.

### Something to think about

3. **Cross-account update/patch tests only cover auth and registration.** `MCPUpdateCrossAccountTests` has no success-path test for either `update_cross_account_request` or `patch_cross_account_request` -- only no-auth and config registration checks. This matches the existing pattern for `MCPCreateCrossAccountTests` (cross-account setup is complex), so it's consistent. Just wanted to flag that the actual view delegation path isn't exercised for these two tools.

Looks good otherwise. The Sourcery round-1 refactoring was well done and the patterns are consistent across all 15 new tools.

File: rbac/management/mcp_views.py [hunk 1]
```diff
@@ -280,29 +305,62 @@ def _clone_request(
...
+    if query_params:
+        from urllib.parse import urlencode
+
+        path = f"{path}?{urlencode(query_params)}"
+    view_request = _clone_request(request, path, method=method, body=body)
     response = view(view_request, **view_kwargs)
     if hasattr(response, "render"):
         response = response.render()
+    if response.status_code == 204:
+        return json.dumps({"status": "deleted"})
```
[the remark this answers is not in the document]
Comment (dsova06) [reply rbac/management/mcp_views.py]: Fixed: changed 204 response from `{"status": "deleted"}` to `{"status": "no_content"}` in `_call_view_json`. Updated all 6 test assertions to match.

File: tests/management/test_mcp_views.py [hunk 2]
```diff
@@ -4611,3 +4626,666 @@ def test_create_cross_account_request_success(self):
...
+class MCPUpdateCrossAccountTests(MCPToolTestMixin, IdentityRequest):
+    """Test update_cross_account_request and patch_cross_account_request."""
+
+    def setUp(self):
+        """Set up cross-account update tests."""
+        super().setUp()
+        self.url = "/_private/_a2s/mcp/"
+        self.client = APIClient()
+        self.principal = Principal.objects.create(username="test_user", tenant=self.tenant)
+
```
[the remark this answers is not in the document]
Comment (dsova06) [reply tests/management/test_mcp_views.py:4963]: Fixed: added `test_patch_cross_account_request_rejects_invalid_status` that calls with `status="foo"` and asserts the error mentions the invalid value and the allowed statuses.

**Comment by dsova06** [comment]
@sourcery-ai resolve

**Review by lpichler — APPROVED** [review]

**Comment by lpichler** [comment]
@dsova06 just linter:

```
py312: OK ✔ in 2 minutes 41.94 seconds
mypy: install_deps> python -I -m pip install mypy types-python-dateutil
mypy: commands[0]> mypy --install-types --non-interactive rbac
rbac/management/mcp_views.py:350: error: Argument 6 to "_call_view_json" has incompatible type "**dict[str, str]"; expected "dict[str, str] | None"  [arg-type]
Found 2 errors in 1 file (checked 334 source files)
rbac/management/mcp_views.py:350: note: "dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
rbac/management/mcp_views.py:350: note: Consider using "Mapping" instead, which is covariant in the value type
rbac/management/mcp_views.py:361: error: Argument 6 to "_call_view_json" has incompatible type "**dict[str, str]"; expected "dict[str, Any] | None"  [arg-type]
rbac/management/mcp_views.py:361: note: "dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
rbac/management/mcp_views.py:361: note: Consider using "Mapping" instead, which is covariant in the value type
mypy: exit 1 (3.26 seconds) /var/workdir> mypy --install-types --non-interactive rbac pid=3654
  lint: OK (7.92=setup[4.30]+cmd[0.97,2.65] seconds)
  py312: OK (161.94=setup[4.70]+cmd[28.09,102.73,6.34,20.08] seconds)
  mypy: FAIL code 1 (6.08=setup[2.82]+cmd[3.26] seconds)
  evaluation failed :( (175.96 seconds)
step-use-trusted-artifact :-
```
